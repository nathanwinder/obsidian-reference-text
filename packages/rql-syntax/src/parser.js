// This file was generated by lezer-generator. You probably shouldn't edit it.
import {LRParser} from "@lezer/lr"
import {aliasTextOrSymbol} from "./tokens"
export const parser = LRParser.deserialize({
  version: 14,
  states: ")xQfQQOOO!oQQO'#DQO!yQQO'#C`O#RQQO'#DPO#YOQO'#CeO#bOPO'#DUOOQO'#Cf'#CfOOQO'#DP'#DPOOQO'#Cn'#CnO#xQQO'#CkO#gQQO'#C}O$]QQO'#C^O$bQQO'#CjO$sQQO'#CiOOQO'#Cm'#CmQfQQOOO%RQQO,59OO%RQQO'#CoO%^QQO,58zOOQO,59P,59POOQO,59p,59pOOQO-E6l-E6lO&UQQO'#C_O&`QQO,59iO&hOPO'#ChO&mQQO,58xOOQO'#Cl'#ClOOQO'#Cr'#CrO'XQQO,59UOOQO'#Cq'#CqO'jQQO,59TOfQQO,59TO(UQQO,59TO'jQQO,59TOOQO,59T,59TOOQO-E6k-E6kOOQO'#DQ'#DQOOQO1G.j1G.jO(dQQO,59ZOOQO,59Z,59ZOOQO-E6m-E6mO#gQQO'#CpO)_QQO1G/TOOQO1G/T1G/TOOQO,59S,59SO)gQQO1G.dOOQO-E6p-E6pOOQO-E6o-E6oO*RQQO1G.oOOQO,59_,59_O*aQQO1G.oOfQQO1G.oOOQO-E6q-E6qO*aQQO1G.oOOQO1G.o1G.oO*{QQO'#CgOOQO,59[,59[OOQO-E6n-E6nOOQO7+$o7+$oO+VQQO7+$ZO+VQQO7+$ZOOQO7+$Z7+$ZO+qQQO7+$ZP#gQQO'#CsO,PQQO<<GuO,PQQO<<GuOOQO<<Gu<<GuO,kQQOAN=a",
  stateData: "-Y~OkOSlOSmOSnOSoOSpOS~OTPOUPOVPOrYOwSOyTO~OTsXUsXVsXrsXwsXysXzsX|sX}sX!OsX{sX~OutXvtX~PzOu`OvaO~OvaO~PzOTcOUcO~OidO~OTPOUPOVPOwSOyTO~Or_Xz_X|_X}_X!O_X~P#gO|hO~OrYOz^X|^X}^X!O^X~OzoO|hO}mO!OrO~OTtOUtOVtO~OvaOTSaUSaVSarSawSaySazSa|Sa}Sa!OSa{Sa~OzRX{RX~P#gOzyO{{O~Oi|O~O}mOTQaUQaVQahQarQawQayQa~OrYOz^a|^a}^a!O^a~O}mOT]aU]aV]ah]ar]aw]ay]a~Oz!TO|hO}mO!O!WO~Ou`OTcaUcaVcarcavcawcaycazca|ca}ca!Oca{ca~OzyO{![O~O}mOTQiUQiVQihQirQiwQiyQi~OrYO|hO}mO!O!_O~O}mOT]iU]iV]ih]ir]iw]iy]i~OzZX{ZX~P#gO}mOT]qU]qV]qh]qr]qw]qy]q~OrYO|hO}mO!O!dO~O}mOT]yU]yV]yh]yr]yw]yy]y~O}mOT]!RU]!RV]!Rh]!Rr]!Rw]!Ry]!R~O}k~",
  goto: "&lyPPz!O!RPPP!_!R!R!n!qz#Q#Z#b#j#p$S$Z$a%O%^PPPPPPPPP%dP%p%|PPP&`T^O_RgYeVOXY_foy!T!X!adROXY_foy!T!X!aRwaR!YyQiZQq]Q!VpQ!^!QR!c!`S]O_V!Ro!T!aZ[O_o!T!a]k[lo!Q!T!`Q_ORs_YXO_o!T!aUeXf!XQfYR!XySbQRRxbQzgR!ZzQn]Q}i[!Pn}!S!]!b!eS!SpqS!]!Q!VS!b!^!`R!e!cQl[U!Ol!Q!`Q!QoR!`!TQp]R!UpSZO_]j[lo!Q!T!`eWOXY_foy!T!X!adQOXY_foy!T!X!aQu`RvaeUOXY_foy!T!X!a",
  nodeNames: "âš  Document Transform Operation Range Symbol Text Placeholder Span Tag Alias Operation Alias Block Reference Select Transform",
  maxTerm: 46,
  skippedNodes: [0],
  repeatNodeCount: 7,
  tokenData: ")`~RgXY!jYZ!u]^#_pq!jrs#wst%ovw%|wx&Rxy't|}'y}!O(O!O!P(T!P!Q(Y!Q![(_![!](p!a!b(u!c!}(_!}#O(z#O#P)P#P#Q)U#R#S(_#T#o(_~~)Z~!oQl~XY!jpq!j~!zQk~YZ#Q]^#V~#VO}~~#[P}~YZ#Q~#dQk~YZ#j]^#V~#qQ}~k~YZ#Q]^#V~#|WU~OY#wZr#wrs$fs#O#w#O#P$k#P;'S#w;'S;=`%i<%lO#w~$kOU~~$nRO;'S#w;'S;=`$w;=`O#w~$|XU~OY#wZr#wrs$fs#O#w#O#P$k#P;'S#w;'S;=`%i;=`<%l#w<%lO#w~%lP;=`<%l#w~%tPw~xy%w~%|Oy~~&ROz~~&WWU~OY&RZw&Rwx$fx#O&R#O#P&p#P;'S&R;'S;=`'n<%lO&R~&sRO;'S&R;'S;=`&|;=`O&R~'RXU~OY&RZw&Rwx$fx#O&R#O#P&p#P;'S&R;'S;=`'n;=`<%l&R<%lO&R~'qP;=`<%l&R~'yO|~~(OOv~~(TOu~~(YOn~~(_Oo~~(dST~!Q![(_!c!}(_#R#S(_#T#o(_~(uOm~~(zOV~~)POr~~)UOp~~)ZO{~~)`O!O~",
  tokenizers: [aliasTextOrSymbol, 0],
  topRules: {"Document":[0,1]},
  tokenPrec: 559
})
