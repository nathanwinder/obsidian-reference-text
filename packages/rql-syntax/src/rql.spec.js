import {parser} from "./parser.js";

describe("valid rql", ()=> {
    it("single symbol flag", test(`#flag`, e=> e.toMatchSnapshot()))

    it("single text flag", test(`#"flag"`, e=> e.toMatchSnapshot()))

    it("single flag", test(`#flag`, e=> e.toMatchSnapshot()))

    it("source and flag", test(`source #flag`, e=> e.toMatchSnapshot()))

    it("source and symbol tag", test(`source #tag value`, e=> e.toMatchSnapshot()))

    it("source and text tag", test(`source #tag "text value"`, e=> e.toMatchSnapshot()))

    it("symbol span", test(`1-5`, e=> e.toMatchSnapshot()))

    it("text span", test(`"Text 1"-"Text 5"`, e=> e.toMatchSnapshot()))

    it("symbol range", test(`1, 5`, e=> e.toMatchSnapshot()))

    it("text range", test(`"Text 1", "Text 5"`, e=> e.toMatchSnapshot()))

    it("complex range", test(`1-2, 5, "Text 9", "Text 10" - "Text 13"`, e=> e.toMatchSnapshot()))

    it("simple transform", test(`[Highlight red] (HR)`, e=> e.toMatchSnapshot()))

    it("multi-transform", test(`[Highlight red & Underline blue] (Hr Ub)`, e=> e.toMatchSnapshot()))

    it("simple reference", test(`John 3:16`, e=> e.toMatchSnapshot()))

    it("multi-operation reference", test(`John Niv 3:16 & John Esv 3:16 & John Nasb 3:16`, e=> e.toMatchSnapshot()))

    it("reference operation transform", test(`John 3:16 ["God" & Highlight]`, e=> e.toMatchSnapshot()))

    it("reference operation multi-transform", test(`John 3:16 ["God" & Highlight red & Underline blue]`, e=> e.toMatchSnapshot()))

    it("reference multi-operation multi-transform reference-transform", test(`John ESV 3:16 [ "God" & H ] & John NIV 3:16 [ "God" & U ] & [ "God" & bold ]`, e=> e.toMatchSnapshot()))

    it("multi-line document", test(`
    [Highlight red] (hr) [Underline blue] (ub)
    [hr & ub] (hrub)

    [strike] (s)
    Bibles Niv John (John)



    John 3:16-17, 19
    ["God" & hrub & italics]
    ["Son"
    & hrub
    & itaclis]

    James 1:1`, e=> e.toMatchSnapshot()))

});

describe("invalid rql", ()=> {
    it("multiple odd span", test(`1-2-3`, e=> e.toMatchSnapshot()))
    it("multiple even span", test(`1-2-3-4`, e=> e.toMatchSnapshot()))
});

/**
 * 
 * @param {string} expression 
 * @param {} assertion this is done so that test tooling will see "toMatchSnapshot" 
 * call in the test and add the appropriate context menu options to view or update the snapshot. 
 */
function test(expression, assertion){
    return ()=> assertion(expect(render(expression)))
}

/**
 * 
 * @param {string} expression 
 */
function render(expression){
    let tree = parser.parse(expression);
    let output = "";
    let depth = 0;
    tree.iterate({
        enter: (node) => {
            const hasChild = node.node.firstChild != null;
            const value = hasChild? "": `<${expression.substring(node.from, node.to)}>`;
            output += `\n${"Â·".repeat(depth)}${node.name}${value}`; 
            depth++;
        },
        leave: (node) => {
            depth--;
        }
    });
    return "\n" + expression + "\n----------------------------------------------" + output + "\n";
}
