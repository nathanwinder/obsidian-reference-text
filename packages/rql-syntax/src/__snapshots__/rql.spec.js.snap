// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid rql multiple even span 1`] = `
"
1-2-3-4
----------------------------------------------
Document
·Block
··Reference
···Select
····Span
·····Symbol<1>
·····Symbol<2>
····⚠<->
····Span
·····Symbol<3>
·····Symbol<4>
"
`;

exports[`invalid rql multiple odd span 1`] = `
"
1-2-3
----------------------------------------------
Document
·Block
··Reference
···Select
····Span
·····Symbol<1>
·····Symbol<2>
····⚠<->
····Symbol<3>
"
`;

exports[`valid rql complex range 1`] = `
"
1-2, 5, "Text 9", "Text 10" - "Text 13"
----------------------------------------------
Document
·Block
··Reference
···Select
····Range
·····Span
······Symbol<1>
······Symbol<2>
·····Symbol<5>
·····Text<"Text 9">
·····Span
······Text<"Text 10">
······Text<"Text 13">
"
`;

exports[`valid rql multi-line document 1`] = `
"

    [Highlight red] (hr) [Underline blue] (ub)
    [hr & ub] (hrub)

    [strike] (s)
    Bibles Niv John (John)



    John 3:16-17, 19
    ["God" & hrub & italics]
    ["Son"
    & hrub
    & itaclis]

    James 1:1
----------------------------------------------
Document
·Transform
··Operation
···Symbol<Highlight>
···Symbol<red>
··Alias
···Symbol<hr)>
·Transform
··Operation
···Symbol<Underline>
···Symbol<blue>
··Alias
···Symbol<ub)>
·Transform
··Operation
···Symbol<hr>
··Operation
···Symbol<ub>
··Alias
···Symbol<hrub)>
·Transform
··Operation
···Symbol<strike>
··Alias
···Symbol<s)>
·Block
··Reference
···Select
····Symbol<Bibles>
····Symbol<Niv>
····Symbol<John>
··Alias
···Symbol<John)>
·Block
··Reference
···Select
····Symbol<John>
····Symbol<3>
····Range
·····Span
······Symbol<16>
······Symbol<17>
·····Symbol<19>
···Transform
····Operation
·····Text<"God">
····Operation
·····Symbol<hrub>
····Operation
·····Symbol<italics>
···Transform
····Operation
·····Text<"Son">
····Operation
·····Symbol<hrub>
····Operation
·····Symbol<itaclis>
·Block
··Reference
···Select
····Symbol<James>
····Symbol<1>
····Symbol<1>
"
`;

exports[`valid rql multi-operation reference 1`] = `
"
John Niv 3:16 & John Esv 3:16 & John Nasb 3:16
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<John>
····Symbol<Niv>
····Symbol<3>
····Symbol<16>
··Reference
···Select
····Symbol<John>
····Symbol<Esv>
····Symbol<3>
····Symbol<16>
··Reference
···Select
····Symbol<John>
····Symbol<Nasb>
····Symbol<3>
····Symbol<16>
"
`;

exports[`valid rql multi-transform 1`] = `
"
[Highlight red & Underline blue] (Hr Ub)
----------------------------------------------
Document
·Transform
··Operation
···Symbol<Highlight>
···Symbol<red>
··Operation
···Symbol<Underline>
···Symbol<blue>
··Alias
···⚠<>
·Block
··Reference
···Select
····Symbol<Hr>
····Symbol<Ub>
····⚠<)>
"
`;

exports[`valid rql reference multi-operation multi-transform reference-transform 1`] = `
"
John ESV 3:16 [ "God" & H ] & John NIV 3:16 [ "God" & U ] & [ "God" & bold ]
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<John>
····Symbol<ESV>
····Symbol<3>
····Symbol<16>
···Transform
····Operation
·····Text<"God">
····Operation
·····Symbol<H>
··Reference
···Select
····Symbol<John>
····Symbol<NIV>
····Symbol<3>
····Symbol<16>
···Transform
····Operation
·····Text<"God">
····Operation
·····Symbol<U>
··Transform
···Operation
····Text<"God">
···Operation
····Symbol<bold>
"
`;

exports[`valid rql reference operation multi-transform 1`] = `
"
John 3:16 ["God" & Highlight red & Underline blue]
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<John>
····Symbol<3>
····Symbol<16>
···Transform
····Operation
·····Text<"God">
····Operation
·····Symbol<Highlight>
·····Symbol<red>
····Operation
·····Symbol<Underline>
·····Symbol<blue>
"
`;

exports[`valid rql reference operation transform 1`] = `
"
John 3:16 ["God" & Highlight]
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<John>
····Symbol<3>
····Symbol<16>
···Transform
····Operation
·····Text<"God">
····Operation
·····Symbol<Highlight>
"
`;

exports[`valid rql simple reference 1`] = `
"
John 3:16
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<John>
····Symbol<3>
····Symbol<16>
"
`;

exports[`valid rql simple transform 1`] = `
"
[Highlight red] (HR)
----------------------------------------------
Document
·Transform
··Operation
···Symbol<Highlight>
···Symbol<red>
··Alias
···⚠<>
·Block
··Reference
···Select
····Symbol<HR>
····⚠<)>
"
`;

exports[`valid rql single flag 1`] = `
"
#flag
----------------------------------------------
Document
·Block
··Reference
···Select
····Tag
·····Symbol<flag>
"
`;

exports[`valid rql single symbol flag 1`] = `
"
#flag
----------------------------------------------
Document
·Block
··Reference
···Select
····Tag
·····Symbol<flag>
"
`;

exports[`valid rql single text flag 1`] = `
"
#"flag"
----------------------------------------------
Document
·Block
··Reference
···Select
····Tag
·····Text<"flag">
"
`;

exports[`valid rql source and flag 1`] = `
"
source #flag
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<source>
····Tag
·····Symbol<flag>
"
`;

exports[`valid rql source and symbol tag 1`] = `
"
source #tag value
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<source>
····Tag
·····Symbol<tag>
····Symbol<value>
"
`;

exports[`valid rql source and text tag 1`] = `
"
source #tag "text value"
----------------------------------------------
Document
·Block
··Reference
···Select
····Symbol<source>
····Tag
·····Symbol<tag>
····Text<"text value">
"
`;

exports[`valid rql symbol range 1`] = `
"
1, 5
----------------------------------------------
Document
·Block
··Reference
···Select
····Range
·····Symbol<1>
·····Symbol<5>
"
`;

exports[`valid rql symbol span 1`] = `
"
1-5
----------------------------------------------
Document
·Block
··Reference
···Select
····Span
·····Symbol<1>
·····Symbol<5>
"
`;

exports[`valid rql text range 1`] = `
"
"Text 1", "Text 5"
----------------------------------------------
Document
·Block
··Reference
···Select
····Range
·····Text<"Text 1">
·····Text<"Text 5">
"
`;

exports[`valid rql text span 1`] = `
"
"Text 1"-"Text 5"
----------------------------------------------
Document
·Block
··Reference
···Select
····Span
·····Text<"Text 1">
·····Text<"Text 5">
"
`;
